### PR TITLE
Support unknown rank image

### LINKED ISSUE #332: Check which image ops are compatible with tf.data.Dataset
Follow up https://github.com/tensorflow/addons/issues/243.

Here is a check list:
+ [x] dense_image_warp
+ [ ] sparse_image_warp		
+ [x] interpolate_bilinear	
+ [x] euclidean_distance_transform	
+ [x] adjust_hsv_in_yiq	
+ [x] random_hsv_in_yiq	
+ [x] mean_filter2d	
+ [x] median_filter2d	
+ [x] angles_to_projective_transforms	
+ [x] matrices_to_flat_transforms	
+ [x] rotate	
+ [x] transform

Like what we did in https://github.com/tensorflow/addons/pull/330, feel free to open a pull request for each op:
1. Add test for unknown rank image tensor.
2. Fix it if possible. Otherwise, please notice users that the op is not compatible with unknown rank tensor (shape=None) in its document.

### PR DESCRIPTION
**Pull request by WindQAQ**
# Description

Resubmit of #446 with unittests. Closes #332.

## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [x] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

Passes existing tests and new tests.

### CODE REVIEW

File: tensorflow_addons/image/sparse_image_warp.py [hunk 1]
```diff
@@ -140,33 +146,38 @@ def sparse_image_warp(
...
-      tf.float64, and do not necessarily have to be the same type.
+      Note that `image` and `offsets` can be of type `tf.half`, `tf.float32`, or
+      `tf.float64`, and do not necessarily have to be the same type.
 
     Returns:
-      warped_image: `[batch, height, width, channels]` float `Tensor` with same
-        type as input image.
-      flow_field: `[batch, height, width, 2]` float `Tensor` containing the
+      warped_image: a float `Tensor` with the same shape and dtype as `image`.
+      flow_field: `[batch_size, height, width, 2]` float `Tensor` containing the
```
Comment (WindQAQ) [inline tensorflow_addons/image/sparse_image_warp.py:173]: It's better to squeeze the batch dim of `flow_field` when `image` is 2-D or 3-D. The `batch_size` is assumed to be 1 as documentation above. We can deal with it if users request as most of `sparse_image_warp` users want the first output only.

**Review by seanpmorgan — APPROVED** [review]
LGTM Thanks!

**Comment by lutzzz** [comment]
@WindQAQ Thanks for your great work for removing `numpy` from the code, but not all the problems mentioned in #446 and #332 regarding `tf.data.Dataset` were solved by this PR. 
It is still not possible to use `sparse_image_warp` in `Dataset.map`, where the function is called with tensors of known rank but dynamic/unknown shape (e.g., (None, 32, 32, 3)). Thus, running your code raises an error

```ValueError: Tried to convert 'input' to a tensor and failed. Error: None values not supported.```

in `_expand_to_minibatch:37`

```batch_size = tf.expand_dims(batch_size, 0)```

Could you improve this, please?

**Comment by WindQAQ** [comment]
> @WindQAQ Thanks for your great work for removing `numpy` from the code, but not all the problems mentioned in #446 and #332 regarding `tf.data.Dataset` were solved by this PR.
> It is still not possible to use `sparse_image_warp` in `Dataset.map`, where the function is called with tensors of known rank but dynamic/unknown shape (e.g., (None, 32, 32, 3)). Thus, running your code raises an error
> 
> `ValueError: Tried to convert 'input' to a tensor and failed. Error: None values not supported.`
> 
> in `_expand_to_minibatch:37`
> 
> `batch_size = tf.expand_dims(batch_size, 0)`
> 
> Could you improve this, please?

Hi @lutzzz, thanks for reporting. Can you provide the minimal code snippet to reproduce it? I can not manipulate the issue with `get_concrete_function`.

**Comment by WindQAQ** [comment]
NVM. I find the issue. Will ship a PR soon.

**Pull request merged**